qiskit_experiments/data_processing/data_processor.py

```diff
@@ -125,79 +190,117 @@ def _call_internal(
+                ``self._input_key``.
 
         Returns:
             The data formatted in such a way that it is ready to be processed by the nodes.
 
         Raises:
             DataProcessorError:
                 - If the input datum is not a list or a dict.
-                - If the data processor received a single datum but requires all the data to
-                  process it properly.
                 - If the input key of the data processor is not contained in the data.
+                - If the data processor receives multiple data with different measurement
+                  configuration, i.e. Jagged array.
         """
         if isinstance(data, dict):
             data = [data]
 
+        data_to_process = []
+        dims = None
+        for datum in data:
+            try:
+                outcome = datum[self._input_key]
+            except TypeError as error:
+                raise DataProcessorError(
+                    f"{self.__class__.__name__} only extracts data from "
+                    f"lists or dicts, received {type(data)}."
+                ) from error
+            except KeyError as error:
+                raise DataProcessorError(
+                    f"The input key {self._input_key} was not found in the input datum."
+                ) from error
+
+            if self._input_key != "counts":
+                outcome = np.asarray(outcome)
+                # Validate data shape
+                if dims is None:
+                    dims = outcome.shape
+                else:
+                    # This is because each data node creates full array of all result data.
+                    # Jagged array cannot be numerically operated with numpy array.
+                    if outcome.shape != dims:
+                        raise DataProcessorError(
+                            "Input data is likely a mixture of job results with different "
+                            "measurement setup. Data processor doesn't support jagged array."
+                        )
+            data_to_process.append(outcome)
+
         try:
-            data_ = [_datum[self._input_key] for _datum in iter(data)]
-        except KeyError as error:
-            raise DataProcessorError(
-                f"The input key {self._input_key} was not found in the input datum."
-            ) from error
-        except TypeError as error:
-            raise DataProcessorError(
-                f"{self.__class__.__name__} only extracts data from "
-                f"lists or dicts, received {type(data)}."
-            ) from error
-
-        return data_
+            # Likely level1 or below. Return ufloat array with un-computed std_dev.
+            # The output data format is a standard ndarray with dtype=object with
+            # arbitrary shape [n_circuits, ...] depending on the measurement setup.
+            nominal_values = np.asarray(data_to_process, float)
+            return unp.uarray(
```

Comment:
With unumpy module, we can initialize the processed data as a single numpy array instead of tuple of list and `None`. Thus we don't need to typecast data to array (`asarray`), and frequently appearing code in the format method
```python
if error is not None:
   ...
```
can be removed too. This means we are no longer need `DataAction._format` logic, and now it is almost looking like `_validate` method. If we can rename the method, the branching `if self._validate:` can be moved to `__call__` method, and we can further simplify the logic.